### PR TITLE
[FLINK-12501] Use SpecificRecord.getSchema in AvroFactory

### PR DESCRIPTION
**Pull request by aljoscha**
## What is the purpose of the change

Before, we were using SpecificData.getSchema(type) which was not working
for types that were generated using Avrohugger (for Scala) because
the SCHEMA was generated in the companion object. Now we use a method
that must be available on all SpecificRecord(s).

We still use the old method as a fallback if we cannot instantiate or
call getSchema() on the instance.

## Brief change log

  - We create an instance of the `SpecificRecord` using `Class.newInstance` and then call `SpecificRecord.getSchema()`. We use the old method as a fallback

## Verifying this change

This is covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - The serializers: yes, the avro serializer

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by aljoscha** [comment]
R: @igalshilman could you maybe review this?

File: flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroFactory.java [hunk 1]
```diff
@@ -94,7 +94,8 @@ static Schema parseSchemaString(@Nullable String schemaString) {
 	@SuppressWarnings("OptionalUsedAsFieldOrParameterType")
 	private static <T> AvroFactory<T> fromSpecific(Class<T> type, ClassLoader cl, Optional<Schema> previousSchema) {
 		SpecificData specificData = new SpecificData(cl);
-		Schema newSchema = specificData.getSchema(type);
+		Optional<Schema> newSchemaOptional = tryExtractAvroSchema(type);
+		Schema newSchema = newSchemaOptional.orElse(specificData.getSchema(type));
```
Comment (debasishg) [inline flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroFactory.java]: I faced a problem here which comes from the use of `Optional.orElse`. It's a known gotcha with `orElse`. Even though `newSchemaOptional` is non-empty, `orElse` will evaluate the argument passed as default. Hence it tries to evaluate `specificData.getSchema(type))` and fails with the earlier exception. The workaround is to change `orElse` to `orElseGet` and pass a lambda instead ..

```java
Schema newSchema = newSchemaOptional.orElseGet(() -> specificData.getSchema(type));
```

This fixes the problem and runs ok. There are a few other invocations of `orElse` in this class which you may want to change to `orElseGet` as well ..

I cherrypicked your commit into the `release-1.9` branch and did a build which runs fine. Thanks for the help.

Comment (aljoscha) [reply flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroFactory.java]: Thanks! That's very helpful. I'll update the PR.

**Comment by aljoscha** [comment]
I updated the PR.

**Comment by debasishg** [comment]
Thanks a lot!

**Comment by aljoscha** [comment]
@flinkbot run travis

**Comment by aljoscha** [comment]
@flinkbot run travis

**Comment by aljoscha** [comment]
@debasishg We did some changes because I forgot to also adapt the `AvroSerializerSnapshot`. Could you again check if this works with your Avrohugger classes? Maybe also check whether you can successfully take a savepoint and restore, because that part was not previously covered by the changes.

**Comment by debasishg** [comment]
@aljoscha Sure I will check all of them ..

**Comment by debasishg** [comment]
works ok .. Thanks!

**Comment by aljoscha** [comment]
@debasishg thanks again! I merged this now.

**Comment by rsheldiii** [comment]
would a similar wrapper around [AvroDeserializationSchema's checkAvroInitialized](https://github.com/apache/flink/blob/master/flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroDeserializationSchema.java#L145) function allow for avro-hugger deserialization support? https://github.com/julianpeeters/avrohugger/issues/111

**Comment by aljoscha** [comment]
We could probably do the same as in https://github.com/apache/flink/pull/9565

**Pull request closed without merging**